Check and apply max deduction for handgrading annotations

broad-well opened the pull request:
This PR fixes #502. When an annotation is applied or deleted in the Handgrading interface, we compute the effective change to the score based on its max deduction and total number of applications so far.

Tests are included for both applying and deleting annotations across the max deduction boundary. Additional tests are included for cases where the max deduction of an annotation is not a multiple of its unit deduction, such as ("-3/-7 max"). This PR's implementation changes makes those tests pass.

## Screenshots

**After**

![Screenshot 2024-04-20 at 12-34-51 Autograder io](https://github.com/eecs-autograder/ag-website-vue/assets/10491561/90547081-e435-4efa-bf9f-85311dab91f8)

Commit:
Make "Actual exit status" always black when return code is not checked

Commit:
Check max deduction for handgrading annotations

Commit:
Undo changes inherited from other PR

james-perretta commented on tests/test_components/test_project_view/test_handgrading/test_handgrading.ts:
```diff
@@ -637,6 +650,109 @@ describe('Comment tests', () => {
+            );
+            ag_cli.AppliedAnnotation.notify_applied_annotation_created(applied_annotation);
+        }
+        wrapper.vm.$forceUpdate();
```
I tend to avoid `$forceUpdate` to make sure that the page is re-rendered appropriately. Does it work with `expect(await wait_until(wrapper, w => !w.vm.saving)).toBe(true);` instead?

Normanwqn replied:
I will test with that. Is there a reason that `$forceUpdate` is used here?
```
 test('Handgraders allowed to leave custom comments', async () => {
        wrapper.vm.d_handgrading_result!.handgrading_rubric.handgraders_can_leave_comments = true;
        data_ut.set_global_user_roles(data_ut.make_user_roles({is_handgrader: true}));
        wrapper.vm.$forceUpdate();
        await wrapper.vm.$nextTick();
        expect(wrapper.find('#new-comment').exists()).toBe(true);

        expect(wrapper.findAll('.comment').at(0).find('.delete').exists()).toBe(true);
        expect(wrapper.findAll('.comment').at(3).find('.delete').exists()).toBe(true);
        // Can always delete applied annotations
        expect(wrapper.findAll('.comment').at(1).find('.delete').exists()).toBe(true);
        expect(wrapper.findAll('.comment').at(2).find('.delete').exists()).toBe(true);
    });
  ```

Normanwqn committed:
Added comment for modifying return_code_correct

Normanwqn committed:
Replace forceUpdate with checking for w.vm.saving

Normanwqn committed:
Use checking w.vm.saving in test_handgrading

Replaced forceUpdate with expect(await wait_until(wrapper, w => !w.vm.saving)).toBe(true);
previously written test case Handgraders allowed to leave custom
comments

Normanwqn commented:
@james-perretta
We changed `$forceUpdate` to `expect(await wait_until(wrapper, w => !w.vm.saving)).toBe(true);`, and all unit tests are still passed

We also changed a previously written test case `Handgraders allowed to leave custom comments` which uses `$forceUpdate`, to use ` `expect(await wait_until(wrapper, w => !w.vm.saving)).toBe(true);` instead.

Normanwqn committed:
Merge branch 'develop' into handgrading

james-perretta commented:
@broad-well Did github do that automatically (dismissing the stale review)? For some reason the "merge automatically" check isn't going through on my end, and it still says review required after I approved the PR. I'll check back later to see if it resolves itself, otherwise I'll have you try re-opening the PR.

broad-well commented:
> @broad-well Did github do that automatically (dismissing the stale review)? For some reason the "merge automatically" check isn't going through on my end, and it still says review required after I approved the PR. I'll check back later to see if it resolves itself, otherwise I'll have you try re-opening the PR.

Yes, GitHub automatically dismissed your review. It does this because (1) I branched off to create `handgrading` from `develop` in my fork, (2) I started a PR from my `handgrading` branch to the origin's `develop` branch, and (3) a PR from my `develop` branch to the origin's `develop` branch got merged. [Other users have had the same issue](https://github.com/orgs/community/discussions/58535#discussioncomment-6495938).

We can either duplicate this pull request or have you temporarily turn off "Dismiss stale pull request approvals when new commits are pushed".

james-perretta commented:
Got it, go ahead and make a fresh PR please

On Wednesday, April 24, 2024, Michael P. ***@***.***> wrote:

> @broad-well <https://github.com/broad-well> Did github do that
> automatically (dismissing the stale review)? For some reason the "merge
> automatically" check isn't going through on my end, and it still says
> review required after I approved the PR. I'll check back later to see if it
> resolves itself, otherwise I'll have you try re-opening the PR.
>
> Yes, GitHub automatically dismissed your review. It does this because (1)
> I branched off to create handgrading from develop in my fork, (2) I
> started a PR from my handgrading branch to the origin's develop branch,
> and (3) a PR from my develop branch to the origin's develop branch got
> merged. Other users have had the same issue
> <https://github.com/orgs/community/discussions/58535#discussioncomment-6495938>
> .
>
> We can either duplicate this pull request or have you temporarily turn off
> "Dismiss stale pull request approvals when new commits are pushed".
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eecs-autograder/ag-website-vue/pull/512#issuecomment-2075099315>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA7IXMH3WBLN6LEFSNNNL3LY667HRAVCNFSM6AAAAABGQTW7K6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANZVGA4TSMZRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Normanwqn commented:
@james-perretta  We have submitted another PR <a href="https://github.com/eecs-autograder/ag-website-vue/pull/514"> here </a>. We wold appreciate it if you may take a look. Thank you so much!
